### PR TITLE
feat: add Support for dragging in the trace panel

### PR DESCRIPTION
**Pull request by CodePrometheus**
### Summary

Good morning everyone, the main modification this time is to introduce the drag function in the trace panel, and reference to  https://github.com/apache/skywalking/issues/12004

Video

https://github.com/apache/skywalking-booster-ui/assets/66550292/066d8ee5-e548-4168-9ad8-68c2cd74139d

image

image

Signed-off-by: Zixin Zhou [codeprince2020@163.com](mailto:codeprince2020@163.com)

### CODE REVIEW

**Comment by CodePrometheus** [comment]
@wu-sheng Hello Teacher Wu, regarding the discussion on the automatic expansion of the left panel when searching, I have been thinking about it in the past few weeks. Since expansion and hiding are manually triggered by the user and are strongly related to the direction pointed by the arrow, my idea is that the system does not actively expand and is completely controlled by the user. WDYT? Could I know more about your views, happy to discuss with you.

**Comment by wu-sheng** [comment]
image

The tips are not destroyed.

And, could you verify how this works when the segment list is collapsed and click the new query?

File: src/views/dashboard/related/trace/components/D3Graph/Index.vue [hunk 1]
```diff
@@ -60,6 +61,9 @@ limitations under the License. -->
     }
     draw();
     loading.value = false;
+    emitter.on("trace_resize", () => {
```
Comment (Fine0830) [inline src/views/dashboard/related/trace/components/D3Graph/Index.vue:64]: You should use the [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver/MutationObserver) to observe `div` change, and use `debounceFunc` to draw trace graph.

Comment (Fine0830) [reply src/views/dashboard/related/trace/components/D3Graph/Index.vue:64]: Please remove the `emitter`.

File: src/views/dashboard/related/trace/TraceList.vue [hunk 2]
```diff
@@ -187,7 +188,7 @@ limitations under the License. -->
   }
 
   .list {
-    width: 280px;
+    width: 100%;
```
Comment (Fine0830) [inline src/views/dashboard/related/trace/TraceList.vue:191]: Please set the `min-width` here.

**Review by Fine0830 — CHANGES_REQUESTED** [review]
Thanks @CodePrometheus. Leaving couple of comments.

File: src/views/dashboard/related/trace/components/D3Graph/Index.vue (hunk 1 above)
Comment (CodePrometheus) [reply src/views/dashboard/related/trace/components/D3Graph/Index.vue]: Thank you so much, this is a very convenient way to do dom monitoring, learned.

File: src/views/dashboard/related/trace/TraceList.vue (hunk 2 above)
Comment (CodePrometheus) [reply src/views/dashboard/related/trace/TraceList.vue:192]: Added, and I modified the default width to 280.

**Comment by CodePrometheus** [comment]
> The tips are not destroyed.

The reason is that when re-rendering, I neglected to remove `.d3-tip`, has been corrected.

> And, could you verify how this works when the segment list is collapsed and click the new query?

In the following video, please ignore the circle around the mouse, it's my mac's setting.


https://github.com/apache/skywalking-booster-ui/assets/66550292/8bea8728-e60e-48f0-9da1-4a5d554e7cf0

**Comment by wu-sheng** [comment]
I think when you apply a new query, the segment list sidebar should be pop up automatically. Otherwise, it is confusing, especially when query without trace ID, it would be a long segment list with various traces. If you don't pop up the segment list, but update a specific trace view, the user may thought there is only one trace, which is most likely not the case.

**Comment by CodePrometheus** [comment]
> I think when you apply a new query, the segment list sidebar should be pop up automatically. Otherwise, it is confusing, especially when query without trace ID, it would be a long segment list with various traces. If you don't pop up the segment list, but update a specific trace view, the user may thought there is only one trace, which is most likely not the case.

Got your point, relevant logic has been adjusted.

cc @Fine0830 In order to avoid repeated rendering of the trace with the search event, I used `mitt` to identify the event name, WDYT? have any other suggestions, please let me know, thx.


https://github.com/apache/skywalking-booster-ui/assets/66550292/55952143-0f36-4e13-aa74-81fe74445fd9

**Comment by wu-sheng** [comment]
Logically, this is good for me. Wait for @Fine0830 to check code details.

**Comment by Fine0830** [comment]
> cc @Fine0830 In order to avoid repeated rendering of the trace with the search event, I used `mitt` to identify the event name, WDYT? have any other suggestions, please let me know, thx.
> 
 What do you mean? Why does this feature affect search?

**Comment by CodePrometheus** [comment]
> What do you mean? Why does this feature affect search?

When click search, the trace will be rendered once, and when the segment list reaches the default position, because the width change will also trigger a trace rendering, I think it is only rendered once, so I use the event name to judge.

**Comment by Fine0830** [comment]
> > What do you mean? Why does this feature affect search?
> 
> When click search, the trace will be rendered once, and when the segment list reaches the default position, because the width change will also trigger a trace rendering, I think it is only rendered once, so I use the event name to judge.

Can you remove the div event listener when clicking search and add it back after the trace rendering?

**Comment by CodePrometheus** [comment]
> Can you remove the div event listener when clicking search and add it back after the trace rendering?

I checked and found that the current implementation and what you said will cause the trace-list to be unable to readjust the width after the segment-list rebounds. Is it possible to trigger `draw` in the `watch`?

**Comment by Fine0830** [comment]
> > Can you remove the div event listener when clicking search and add it back after the trace rendering?
> 
> I checked and found that the current implementation and what you said will cause the trace-list to be unable to readjust the width after the segment-list rebounds. Is it possible to trigger `draw` in the `watch`?

We already have it [here](https://github.com/apache/skywalking-booster-ui/blob/main/src/views/dashboard/related/trace/components/D3Graph/Index.vue#L299)

**Comment by CodePrometheus** [comment]
> We already have it [here](https://github.com/apache/skywalking-booster-ui/blob/main/src/views/dashboard/related/trace/components/D3Graph/Index.vue#L299)

Yes, but what I mean is that the original implementation will not adjust the width of the trace-list after segment pop back, so I modified it to trigger function `draw()` directly.

**Comment by wu-sheng** [comment]
Have you committed the codes? @Fine0830 is able to verify the modified codes?

**Comment by Fine0830** [comment]
> > We already have it [here](https://github.com/apache/skywalking-booster-ui/blob/main/src/views/dashboard/related/trace/components/D3Graph/Index.vue#L299)
> 
> Yes, but what I mean is that the original implementation will not adjust the width of the trace-list after segment pop back, so I modified it to trigger function `draw()` directly, [diff](https://github.com/apache/skywalking-booster-ui/pull/377/commits/3c60a7845a31dc446207ca6c7087bebb3b4350e7).

This is okay for me.

**Comment by CodePrometheus** [comment]
Could we proceed to the next step, do I need to provide video for double check?

**Comment by wu-sheng** [comment]
@Fine0830 Could you verify this?

**Comment by Fine0830** [comment]
> Can you remove the div event listener when clicking search and add it back after the trace rendering?

Please change the codes base on this and remove the `mitt`.

**Comment by Fine0830** [comment]
> Can you remove the div event listener when clicking search and add it back after the trace rendering?

You can use [disconnect](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver/disconnect) to stops the MutationObserver instance from receiving further notifications until and unless [observe()](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver/observe) is called again.

**Comment by wu-sheng** [comment]
@CodePrometheus Is the new way working?

**Comment by CodePrometheus** [comment]
> @CodePrometheus Is the new way working?

I've been a little busy at work recently, don't have enough time to look carefully. This way should be achievable, please give me some more time.

**Comment by CodePrometheus** [comment]
The idea of implementation is:
Maintain a Map to represent the relationship between the event name and the corresponding MutationObserver. When entering the page, create the object, disconnect and delete the Key when leaving.
Monitor when the dividing line is dragged and the arrow is clicked, and disconnect when the mouseup occurs.

**Comment by wu-sheng** [comment]
@Fine0830 Please take a look.

**Review by Fine0830 — COMMENTED** [review]
Please remove `mitt.ts`. The Other code is fine.

**Comment by wu-sheng** [comment]
@CodePrometheus Pleaae sync submodule to make this works.

**Pull request merged**